### PR TITLE
Cache Skipped Slots

### PR DESCRIPTION
**Pull request by nisdas**
- [x] Clone State when retrieving from the cache, so that the cached state isnt mutated
- [x] Add regression test to make sure the cache isnt leading to any mutated states

### CODE REVIEW

File: beacon-chain/core/state/transition.go [hunk 1]
```diff
@@ -245,9 +245,35 @@ func ProcessSlots(ctx context.Context, state *pb.BeaconState, slot uint64) (*pb.
...
+			skipSlotCacheHit.Inc()
+		} else {
+			skipSlotCacheMiss.Inc()
+		}
+	}
+
 	for state.Slot < slot {
 		if ctx.Err() != nil {
 			traceutil.AnnotateError(span, ctx.Err())
+			if featureconfig.Get().EnableSkipSlotsCache {
```
Comment (terencechain) [inline beacon-chain/core/state/transition.go]: I dont think you need this duplicated block.   L302 to L307 will always cover this case

File: shared/featureconfig/flags.go [hunk 2]
```diff
@@ -69,10 +69,16 @@ var (
 		Name:  "enable-finalized-block-root-index",
 		Usage: "Enable tracking finalized block roots in database index.",
 	}
+	// EnableSkipSlotsCache enables the skips slots lru cache to be used in runtime.
+	EnableSkipSlotsCache = cli.BoolFlag{
```
Comment (prestonvanloon) [inline shared/featureconfig/flags.go]: This doesnt need to be public

File: beacon-chain/core/state/transition.go (hunk 1 above)
Comment (nisdas) [reply beacon-chain/core/state/transition.go]: what about exceeded deadlines, wouldn't it not be captured then ?

Comment (terencechain) [reply beacon-chain/core/state/transition.go]: Oh i see. That's right

**Pull request merged**